contract/r/gnoswap/gov/governance/upgrade_test.gno

```diff
@@ -0,0 +1,686 @@
+package governance
+
+import (
+	"chain/runtime"
+	"testing"
+
+	"gno.land/p/nt/uassert"
+)
+
+// NOTE: UpgradeImpl tests cannot be included because UpgradeImpl(packagePath string)
+// lacks a `cur realm` parameter, which causes "frame not found" errors when
+// runtime.PreviousRealm() is called in tests. However, proxy and getter function tests
+// CAN be included because RegisterInitializer internally calls updateImplementation(),
+// making the implementation available immediately after RegisterInitializer.
+
+func TestRegisterInitializer(t *testing.T) {
+	tests := []struct {
+		name                 string
+		setup                func(t *testing.T)
+		initializer          func(s IGovernanceStore) IGovernance
+		callerRealm          runtime.Realm
+		expectedHasAbort     bool
+		expectedAbortMessage string
+	}{
+		{
+			name:        "register initializer is success",
+			callerRealm: testing.NewCodeRealm("gno.land/r/gnoswap/gov/governance/v1"),
+			initializer: makeMockInitializer("v1"),
+		},
+		{
+			name: "register initializer is failed by duplicate registration",
+			setup: func(t *testing.T) {
+				testing.SetRealm(testing.NewCodeRealm("gno.land/r/gnoswap/gov/governance/v1"))
+				RegisterInitializer(cross, makeMockInitializer("v1"))
+			},
+			callerRealm:          testing.NewCodeRealm("gno.land/r/gnoswap/gov/governance/v1"),
+			initializer:          makeMockInitializer("v1"),
+			expectedHasAbort:     true,
+			expectedAbortMessage: "version_manager: initializer already registered",
+		},
+		{
+			name:                 "register initializer is failed by invalid domain path",
+			callerRealm:          testing.NewCodeRealm("gno.land/r/gnoswap/invalid"),
+			initializer:          makeMockInitializer("v1"),
+			expectedHasAbort:     true,
+			expectedAbortMessage: "version_manager: caller is not in the domain path",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			resetTestState(t)
+
+			if tt.setup != nil {
+				tt.setup(t)
+			}
+
+			testing.SetRealm(tt.callerRealm)
+
+			if tt.expectedHasAbort {
+				uassert.AbortsContains(t, tt.expectedAbortMessage, func() {
+					RegisterInitializer(cross, tt.initializer)
+				})
+			} else {
+				RegisterInitializer(cross, tt.initializer)
+			}
```

Comment:
The test doesn't verify if `RegisterInitializer` actually updated the implementation. I think this test need to cast `getImplementation()` to `*MockGovernance` and check the `Version` or call count to ensure the duplicate registration/update logic is working correctly.